### PR TITLE
[clang-cache] Handle the errors when lookup results from CAS

### PR DESCRIPTION
**Pull request by cachemeifyoucan**

### CODE REVIEW

**Comment by cachemeifyoucan** [comment]
Remark looks like:
```
remark: compile job cache miss due to error: the result object for key 'b0a7baa46c5efb477cf7c1b125eeaf00d496752d633d2563f08011f5dc96582a' does not exist in CAS: '01c375968ceae98802abbf0632ca28acbd0c4b2343a454dacb2cbf9ccac14f7d' [-Rcompile-job-cache-miss]
```

**Comment by akyrtzi** [comment]
I actually had in mind that an error should be reported as a compiler error and fail the compilation, not just emit a remark (remarks are hidden by default).
Later on we can downgrade or provide a way to configure the behavior if we find a compelling reason. What do you think?

**Comment by akyrtzi** [comment]
Also I did not expect a "does not exist in CAS: [hash]" error to be returned by the action cache itself, I assume this is due to having this tight coupling right now with the ObjectStore; but this is orthogonal.

**Comment by cachemeifyoucan** [comment]
> Also I did not expect a "does not exist in CAS: [hash]" error to be returned by the action cache itself, I assume this is due to having this tight coupling right now with the ObjectStore; but this is orthogonal.

Currently it is required to have `ObjectRef` stored in ActionCache, and `get` will fetch `ObjectRef` too. In case of local CAS, when the hash in ActionCache cannot be converted into an ObjectRef, it will return the error. This is kind of the final proposed design so if you think it should be something different, we should iron it out now.

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

File: clang/include/clang/Basic/DiagnosticCASKinds.td [hunk 1]
```diff
@@ -42,8 +42,7 @@ def remark_compile_job_cache_hit : Remark<
   "compile job cache hit for '%0' => '%1'">, InGroup<CompileJobCacheHit>;
 def remark_compile_job_cache_miss : Remark<
   "compile job cache miss for '%0'">, InGroup<CompileJobCacheMiss>;
-def remark_compile_job_cache_miss_result_not_found : Remark<
-  "compile job cache miss for '%0' (result not found: '%1')">,
-  InGroup<CompileJobCacheMiss>;
+def err_compile_job_cache_result_not_found : Error<
```
Comment (akyrtzi) [inline clang/include/clang/Basic/DiagnosticCASKinds.td]: I'd recommend rename to `err_compile_job_cache_failed`, "result not found" could be confused with "the key is missing".

File: clang/test/CAS/fcache-compile-job.c [hunk 2]
```diff
@@ -3,28 +3,38 @@
 //
 // RUN: %clang -cc1 -triple x86_64-apple-macos11 \
 // RUN:   -fcas-path %t/cas -faction-cache-path %t/cache -fcas-fs @%t/casid -fcache-compile-job \
-// RUN:   -Rcompile-job-cache-hit -emit-obj %s -o %t/output.o 2>&1 \
+// RUN:   -Rcompile-job-cache-hit -Rcompile-job-cache-miss -emit-obj %s -o %t/output.o 2>&1 \
```
Comment (akyrtzi) [inline clang/test/CAS/fcache-compile-job.c]: You could use a single `-Rcompile-job-cache` to enable both.

File: clang/tools/driver/cc1_main.cpp [hunk 3]
```diff
@@ -445,20 +445,21 @@ Optional<int> CompileJobCache::tryReplayCachedResult(CompilerInstance &Clang) {
...
+      Optional<int> Status =
+          replayCachedResult(Clang, **Result, /*JustComputedResult=*/false);
+      assert(Status && "Expected a status for a cache hit");
+      return *Status;
+    }
+    Diags.Report(diag::remark_compile_job_cache_miss)
+        << ResultCacheKey->toString();
+  } else
+    Diags.Report(diag::err_compile_job_cache_result_not_found)
+        << toString(Result.takeError());
```
Comment (akyrtzi) [inline clang/tools/driver/cc1_main.cpp:462]: Shouldn't this be followed with `return 1;`?

Comment (cachemeifyoucan) [reply clang/tools/driver/cc1_main.cpp:462]: This is a fatal error but sure we can return 1 just to be clean.

**Review by akyrtzi — APPROVED** [review]
Thank you!

**Pull request merged**